framework/src/modules/chain/chain.js

```diff
@@ -383,6 +394,37 @@ module.exports = class Chain {
 			totalAmount: this.options.constants.TOTAL_AMOUNT,
 			blockSlotWindow: this.options.constants.BLOCK_SLOT_WINDOW,
 		});
+
+		const blockSyncMechanism = new BlockSynchronizationMechanism({
+			storage: this.storage,
+			logger: this.logger,
+			bft: this.bft,
+			slots: this.slots,
+			channel: this.channel,
+			modules: { blocks: this.blocks },
+			activeDelegates: this.options.constants.ACTIVE_DELEGATES,
+		});
+
+		const fastChainSwitchMechanism = new FastChainSwitchingMechanism({
+			storage: this.storage,
+			logger: this.logger,
+			slots: this.slots,
+			dpos: {},
+			activeDelegates: this.options.constants.ACTIVE_DELEGATES,
+		});
+
+		this.synchronizer = new Synchronizer({
```

Comment:
It was exactly like this before as you described. Modified on request of @limiaspasdaniel

Reply:
I think by doing it the way you propose, the single purpose of the Synchronizer gets invalidated. By doing it the way @nazarhussain is doing it and after long discussing on how to do it, this way, Synchronizer is not populated with too coupled logic and coupled dependencies that come from each individual mechanism. From my point of view, Synchronizer shouldn't be aware of the dependencies of the mechanisms registered to it. This makes it independent, separates its concerns from the ones from its registered mechanisms and improves the readability of the code, as one could simply read the bootstrap initialization logic and register and implement a new synchronization mechanism in an easy way.

Reply:
IMO firing an event from Blocks is not incompatible with the current Synchronizer implementation

Reply:
I tend to agree with @limiaspasdaniel If we put all together in Synchronizer then we simply need two functions rather two distinct isolation for different mechanisms.

Reply:
Maybe I am mistaken but for me, if the synchronizer class constructor initializes two synchronization mechanisms and using the same `isValidFor` interface to choose b/w the two strategies then no strong coupling will be introduced. That being said, I am okay with them initialized from the chain directly.
Although, I would still prefer triggering the synchronizer from the blocks module by firing event rather than declaring the `_handleFastChainStrategy` and `_handleSynchronizationStrategy` funcitons within block processing

Reply:
If the Synchronizer initialized the mechanisms within its constructor, it means it needs to have those dependencies available to be passed down to those mechanisms, hence having a strong coupling to them, at least dependency-wise. However, I agree with your last point of sending an event from the `blocks` module.